### ISSUE TITLE
`convert`ing to a `Union` doesn't work

### ISSUE THREAD

**Issue by StevenWhitaker** [JuliaLang/julia#51697]
On Julia 1.9.3:
```julia
julia> a = Union{Int,String}[1]; a[1] = 2.0
ERROR: MethodError: Cannot `convert` an object of type
  Float64 to an object of type
  Union{Int, String}
```

This seems odd to me, because `a = [1]; a[1] = 2.0` works, as does `a = Integer[1]; a[1] = 2.0`. In particular, in this last example, I read this as "`2.0 isa Integer` is `false`, so convert `2.0` into an object `x` for which `x isa Integer` is `true`" (or something like that), and it seems like similar logic should apply when replacing `Integer` with `Union{Int,String}` (or other unions).

In my particular case, I am adding an `InlineStrings.String15` into an array of type `Vector{Union{Int,String}}`, and it would be nice if Julia did the conversion to `String` for me :)

**Comment by oscardssmith** [comment]
The problem with doing this is that `convert`ing to a `Union` type is potentially ambiguous. i.e. what should `convert(Union{Float64, UInt}, 1)` return?

**Comment by StevenWhitaker** [comment]
Right, in general I can see there might be an issue with that, but if working with a `Union` presumably it doesn't matter which type of the union is used.

**Comment by StevenWhitaker** [comment]
You could also argue that `convert(Integer, 1.0)` is ambiguous: should the result be `1` or `Int32(1)` or `true` or ...

**Comment by adienes** [comment]
> `convert(Integer, 1.0)` is ambiguous:

yeah... imo this shouldn't work either

**Comment by jakobnissen** [comment]
I the Integer case there is a clear case for preferring not to convert. I.e. there is a principle that convert(T, x::T) = x.

**Comment by petvana** [comment]
Related to #49888

**Comment by Tortar** [comment]
Also related https://github.com/JuliaLang/julia/issues/53224
